### PR TITLE
[IOTDB-483] switch all timeGenerate to use TimeColumn and make RawDataSetWithValueFilter to multi thread

### PR DESCRIPTION
**Pull request by liutaohua**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/MaxValueAggrResult.java [hunk 1]
```diff
@@ -66,16 +67,17 @@ public void updateResultFromPageData(BatchData dataInThisPage, long bound) {
...
-      IReaderByTimestamp dataReader) throws IOException {
+  public void updateResultUsingTimestamps(TimeColumn timestamps, long bound,
+      IReaderByTimestamp dataReader)
+      throws IOException {
     Comparable<Object> maxVal = null;
-    for (int i = 0; i < length; i++) {
-      Object value = dataReader.getValueInTimestamp(timestamps[i]);
-      if (value == null) {
+    Object[] value = dataReader.getValuesInTimestamps(timestamps, bound);
+    for (int i = value.length - 1; i >= 0; i--) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/MaxValueAggrResult.java]:
```suggestion
    for (Object value : values) {
```

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/MaxValueAggrResult.java [hunk 2]
```diff
@@ -66,16 +67,17 @@ public void updateResultFromPageData(BatchData dataInThisPage, long bound) {
...
-  public void updateResultUsingTimestamps(long[] timestamps, int length,
-      IReaderByTimestamp dataReader) throws IOException {
+  public void updateResultUsingTimestamps(TimeColumn timestamps, long bound,
+      IReaderByTimestamp dataReader)
+      throws IOException {
     Comparable<Object> maxVal = null;
-    for (int i = 0; i < length; i++) {
-      Object value = dataReader.getValueInTimestamp(timestamps[i]);
-      if (value == null) {
+    Object[] value = dataReader.getValuesInTimestamps(timestamps, bound);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/MaxValueAggrResult.java]:
```suggestion
    Object[] values = dataReader.getValuesInTimestamps(timestamps, bound);
```

File: server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java [hunk 3]
```diff
@@ -19,21 +19,32 @@
 package org.apache.iotdb.db.query.dataset;
 
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java]: useless import

File: server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java [hunk 4]
```diff
@@ -73,26 +86,56 @@ protected RowRecord nextWithoutConstraint() throws IOException {
...
+          Object[] values = readerByTimestamp
+              .getValuesInTimestamps(timeColumn.duplicate(), Long.MAX_VALUE);
+          for (Object value : values) {
+            if (value == null) {
+              fields.add(null);
+            } else {
+              fields.add(Field.getField(value, tsDataType));
+            }
+          }
+          values = null;
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java]: This `values` is not used, maybe it could be removed?

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithValueFilterDataSet.java [hunk 5]
```diff
@@ -37,30 +37,21 @@
...
-  /**
-   * if this object has cached timestamp for next group by partition.
-   */
-  private boolean hasCachedTimestamp;
 
+  private TimeColumn timeColumn = new TimeColumn();
   /**
    * group by batch calculation size.
    */
   protected int timeStampFetchSize;
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithValueFilterDataSet.java]: can be private. And also other methods in this class

File: server/src/main/java/org/apache/iotdb/db/query/executor/AggregationExecutor.java [hunk 6]
```diff
@@ -267,22 +271,17 @@ private void aggregateWithValueFilter(List<AggregateResult> aggregateResults,
       TimeGenerator timestampGenerator, List<IReaderByTimestamp> readersOfSelectedSeries)
       throws IOException {
 
-    while (timestampGenerator.hasNext()) {
-
-      // generate timestamps for aggregate
-      long[] timeArray = new long[aggregateFetchSize];
-      int timeArrayLength = 0;
-      for (int cnt = 0; cnt < aggregateFetchSize; cnt++) {
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/query/executor/AggregationExecutor.java]: Since `aggregateFetchSize` is not used, it could be deleted from this class.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java [hunk 7]
```diff
@@ -73,26 +86,56 @@ protected RowRecord nextWithoutConstraint() throws IOException {
...
-          rowRecord.addField(value, dataTypes.get(i));
+    int seriesNum = seriesReaderByTimestampList.size();
+    while (timeGenerator.hasNextTimeColumn()) {
+      final TimeColumn timeColumn = timeGenerator.nextTimeColumn();
+
+      Future<List<Field>>[] futures = new Future[seriesNum];
+      for (int i = 0; i < seriesNum; i++) {
+        final IReaderByTimestamp readerByTimestamp = seriesReaderByTimestampList.get(i);
+        final TSDataType tsDataType = dataTypes.get(i);
+        futures[i] = QueryTaskPoolManager.getInstance().submit(() -> {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java]: remove parallelization, do this in a separate PR and do some experiments to help us decide whether introduce parallelization

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/MaxValueAggrResult.java (hunk 1 above)
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/MaxValueAggrResult.java]: done

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/MaxValueAggrResult.java (hunk 2 above)
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/MaxValueAggrResult.java]: done

File: server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java (hunk 3 above)
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java]: done

File: server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java (hunk 4 above)
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java]: I think this is in order to reduce the pressure of GC, but it looks as if there is no escape. I've deleted it

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithValueFilterDataSet.java (hunk 5 above)
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithValueFilterDataSet.java]: done

File: server/src/main/java/org/apache/iotdb/db/query/executor/AggregationExecutor.java (hunk 6 above)
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/executor/AggregationExecutor.java]: done

File: server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReaderByTimestamp.java [hunk 8]
```diff
@@ -45,14 +46,19 @@ public SeriesReaderByTimestamp(SeriesReader seriesReader) {
   }
 
   @Override
-  public Object[] getValuesInTimestamps(long[] timestamps) throws IOException {
-    Object[] result = new Object[timestamps.length];
+  public Object[] getValuesInTimestamps(TimeColumn timestamps, long bound) throws IOException {
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/reader/series/SeriesReaderByTimestamp.java]: Not sure. maybe in group by

File: server/src/main/java/org/apache/iotdb/db/query/reader/chunk/DiskChunkReaderByTimestamp.java [hunk 9]
```diff
@@ -40,21 +41,25 @@ public DiskChunkReaderByTimestamp(ChunkReaderByTimestamp chunkReaderByTimestamp)
...
+      currentTime = timestamps.currentTime();
+      timestamps.next();
       while (hasNext()) {
         data = next();
         if (data.getMaxTimestamp() > currentTime) {
           result[i] = null;
           break;
         }
-        result[i] = data.getValueInTimestamp(currentTime);
+        result[i] = new Pair(currentTime, data.getValueInTimestamp(currentTime));
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/reader/chunk/DiskChunkReaderByTimestamp.java]: sorry , forgot

File: server/src/main/java/org/apache/iotdb/db/query/executor/AggregationExecutor.java [hunk 10]
```diff
@@ -267,22 +271,17 @@ private void aggregateWithValueFilter(List<AggregateResult> aggregateResults,
...
+      TimeColumn timeColumn = timestampGenerator.nextTimeColumn();
       // cal part of aggregate result
+      int index = timeColumn.currentIndex();
       for (int i = 0; i < readersOfSelectedSeries.size(); i++) {
-        aggregateResults.get(i).updateResultUsingTimestamps(timeArray, timeArrayLength,
-            readersOfSelectedSeries.get(i));
+        AggregateResult result = aggregateResults.get(i);
+        if (!result.isCalculatedAggregationResult()) {
+          timeColumn.resetIndex(index);
+          result.updateResultUsingTimestamps(timeColumn, Integer.MAX_VALUE,
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/executor/AggregationExecutor.java]: done

File: server/src/main/java/org/apache/iotdb/db/query/executor/AggregationExecutor.java [hunk 11]
```diff
@@ -267,22 +271,17 @@ private void aggregateWithValueFilter(List<AggregateResult> aggregateResults,
...
-        if (!timestampGenerator.hasNext()) {
-          break;
-        }
-        timeArray[timeArrayLength++] = timestampGenerator.next();
-      }
-
+    while (timestampGenerator.hasNextTimeColumn()) {
+      TimeColumn timeColumn = timestampGenerator.nextTimeColumn();
       // cal part of aggregate result
+      int index = timeColumn.currentIndex();
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/executor/AggregationExecutor.java]: yes, I rename the method to` position()`, looks more like the JDK provided,and here can be directly set to 0

File: server/src/main/java/org/apache/iotdb/db/query/externalsort/adapter/ByTimestampReaderAdapter.java [hunk 12]
```diff
@@ -40,14 +41,18 @@ public ByTimestampReaderAdapter(IPointReader pointReader) {
   }
 
   @Override
-  public Object[] getValuesInTimestamps(long[] timestamps) throws IOException {
-    Object[] result = new Object[timestamps.length];
+  public Object[] getValuesInTimestamps(TimeColumn timestamps, long bound) throws IOException {
+    Object[] result = new Object[timestamps.size()];
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/externalsort/adapter/ByTimestampReaderAdapter.java]: `bound` represents the stop time, but we don't know how much data there is between the current index and bound time, and how much memory this Object[] should alloc

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/query/dataset/DataSetWithTimeGenerator.java [hunk 13]
```diff
@@ -56,12 +60,33 @@ public DataSetWithTimeGenerator(List<Path> paths, List<Boolean> cached,
...
+      }
+    }
+    return hasCache;
   }
 
   @Override
   protected RowRecord nextWithoutConstraint() throws IOException {
-    long timestamp = timeGenerator.next();
+    hasCache = false;
+    long timestamp = cacheTime;
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/read/query/dataset/DataSetWithTimeGenerator.java]: done

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithValueFilterDataSet.java [hunk 14]
```diff
@@ -116,67 +108,40 @@ protected RowRecord nextWithoutConstraint() throws IOException {
...
-    while (timestampGenerator.hasNext()) {
-      // construct timestamp array
-      timeArrayLength = constructTimeArrayForOneCal(timestampArray, timeArrayLength);
 
+    while (timestampGenerator.hasNextTimeColumn() || timeColumn.hasCurrent()) {
+      if (timeColumn == null || !timeColumn.hasCurrent()) {
+        timeColumn = timestampGenerator.nextTimeColumn();
+        if (timeColumn.currentTime() >= curEndTime) {
+          break;
+        }
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithValueFilterDataSet.java]: done

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/query/timegenerator/node/OrNode.java [hunk 15]
```diff
@@ -55,17 +55,21 @@ public boolean hasNextTimeColumn() throws IOException {
     }
 
     if (hasLeftValue() && !hasRightValue()) {
-      cachedTimeColumn = leftTimeColumn;
+      while (leftTimeColumn.hasCurrent()) {
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/read/query/timegenerator/node/OrNode.java]: To do this, we need to be careful where `timecolumn` is used in the system, because the `timecolumn` returned may not start at 0

File: server/src/main/java/org/apache/iotdb/db/query/reader/chunk/DiskChunkReaderByTimestamp.java [hunk 16]
```diff
@@ -40,21 +41,25 @@ public DiskChunkReaderByTimestamp(ChunkReaderByTimestamp chunkReaderByTimestamp)
   }
 
   @Override
-  public Object[] getValuesInTimestamps(long[] timestamps) throws IOException {
-    Object[] result = new Object[timestamps.length];
+  public Pair[] getValuesInTimestamps(TimeColumn timestamps, long bound) throws IOException {
+    Pair[] result = new Pair[timestamps.size()];
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/reader/chunk/DiskChunkReaderByTimestamp.java]: `values [I] == null `where this value is used, so I don't think there's going to be any bugs that you can imagine

File: server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java (hunk 7 above)
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java]: done

File: server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java [hunk 17]
```diff
@@ -73,26 +77,62 @@ protected RowRecord nextWithoutConstraint() throws IOException {
...
+      flatAndCache(results, timeColumn);
+    }
+    hasCachedRowRecord = !cachedRecords.isEmpty();
+    return hasCachedRowRecord;
+  }
+
+  private void flatAndCache(Object[][] results, TimeColumn timeColumn) {
+    int resultSize = timeColumn.size() - timeColumn.position();
+    for (int i = 0; i < resultSize; i++) {
+      RowRecord rowRecord = new RowRecord(timeColumn.getTimeByIndex(i + timeColumn.position()));
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java]: record the position before, avoid calculate each time

File: server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java [hunk 18]
```diff
@@ -73,26 +77,62 @@ protected RowRecord nextWithoutConstraint() throws IOException {
...
-        hasCachedRowRecord = true;
-        cachedRowRecord = rowRecord;
-        break;
+        cachedRecords.add(rowRecord);
       }
     }
-    return hasCachedRowRecord;
+  }
+
+  private void fillResults(Object[][] results, TimeColumn timeColumn) throws IOException {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java]: combine the fillResults, initReulsts, flatAndCache, cacheRowRecord into hasNextWithoutConstraints

Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java]: done

File: server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java (hunk 17 above)
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java]: flatAndCache was combined

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithValueFilterDataSet.java [hunk 19]
```diff
@@ -91,13 +75,13 @@ protected void initGroupBy(QueryContext context, GroupByPlan groupByPlan)
     }
   }
 
-  protected TimeGenerator getTimeGenerator(IExpression expression, QueryContext context)
+  private TimeGenerator getTimeGenerator(IExpression expression, QueryContext context)
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithValueFilterDataSet.java]: protected
